go/libraries/doltcore/sqle/dolt_schema_diff_table_function.go

```diff
@@ -0,0 +1,420 @@
+		if ds.tableNameExpr != nil {
+			return fmt.Sprintf("DOLT_SCHEMA_DIFF(%s, %s)", ds.dotCommitExpr.String(), ds.tableNameExpr.String())
+		} else {
+			return fmt.Sprintf("DOLT_SCHEMA_DIFF(%s)", ds.dotCommitExpr.String())
+		}
+	}
+	if ds.tableNameExpr != nil {
+		return fmt.Sprintf("DOLT_SCHEMA_DIFF(%s, %s, %s)", ds.fromCommitExpr.String(), ds.toCommitExpr.String(), ds.tableNameExpr.String())
+	} else {
+		return fmt.Sprintf("DOLT_SCHEMA_DIFF(%s, %s)", ds.fromCommitExpr.String(), ds.toCommitExpr.String())
+	}
+}
+
+// Schema implements the sql.Node interface.
+func (ds *SchemaDiffTableFunction) Schema() sql.Schema {
+	return schemaDiffTableSchema
+}
+
+// Children implements the sql.Node interface.
+func (ds *SchemaDiffTableFunction) Children() []sql.Node {
+	return nil
+}
+
+// WithChildren implements the sql.Node interface.
+func (ds *SchemaDiffTableFunction) WithChildren(children ...sql.Node) (sql.Node, error) {
+	if len(children) != 0 {
+		return nil, fmt.Errorf("unexpected children")
+	}
+	return ds, nil
+}
+
+// CheckPrivileges implements the interface sql.Node.
+func (ds *SchemaDiffTableFunction) CheckPrivileges(ctx *sql.Context, opChecker sql.PrivilegedOperationChecker) bool {
+	if ds.tableNameExpr == nil || !types.IsText(ds.tableNameExpr.Type()) {
+		return true
+	}
+
+	_, _, _, tableName, err := ds.evaluateArguments()
+	if err != nil {
+		return false
+	}
+
+	// TODO: Add tests for privilege checking
+	result := opChecker.UserHasPrivileges(ctx,
+		sql.NewPrivilegedOperation(ds.database.Name(), tableName, "", sql.PrivilegeType_Select))
+	return result
+}
+
+// Expressions implements the sql.Expressioner interface.
+func (ds *SchemaDiffTableFunction) Expressions() []sql.Expression {
+	exprs := []sql.Expression{}
+	if ds.dotCommitExpr != nil {
+		exprs = append(exprs, ds.dotCommitExpr)
+	} else {
+		exprs = append(exprs, ds.fromCommitExpr, ds.toCommitExpr)
+	}
+	if ds.tableNameExpr != nil {
+		exprs = append(exprs, ds.tableNameExpr)
+	}
+	return exprs
+}
+
+// WithExpressions implements the sql.Expressioner interface.
+func (ds *SchemaDiffTableFunction) WithExpressions(expression ...sql.Expression) (sql.Node, error) {
+	if len(expression) < 1 || len(expression) > 3 {
+		return nil, sql.ErrInvalidArgumentNumber.New(ds.Name(), "1 to 3", len(expression))
+	}
+
+	for _, expr := range expression {
+		if !expr.Resolved() {
+			return nil, ErrInvalidNonLiteralArgument.New(ds.Name(), expr.String())
+		}
+		// prepared statements resolve functions beforehand, so above check fails
+		if _, ok := expr.(sql.FunctionExpression); ok {
+			return nil, ErrInvalidNonLiteralArgument.New(ds.Name(), expr.String())
+		}
+	}
+
+	newDstf := *ds
+	if strings.Contains(expression[0].String(), "..") {
+		newDstf.dotCommitExpr = expression[0]
+		if len(expression) > 1 {
+			newDstf.tableNameExpr = expression[1]
+		}
+	} else {
+		if len(expression) < 2 {
+			return nil, sql.ErrInvalidArgumentNumber.New(newDstf.Name(), "1 to 3", len(expression))
```

Comment:
This is if the user has one argument that isn't a dot expression, right? I don't think InvalidArgumentNumber is the right error here. More like ErrInvalidArgumentDetails.

Reply:
Thanks for pointing out ErrInvalidArgumentDetails, changing this error to be clearer.